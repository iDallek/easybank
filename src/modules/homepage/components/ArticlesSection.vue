<template>
  <div class="article__section">
    <h1 class="article__section__title">Lastest Article</h1>
    <div class="articles__area">
      <div v-for="article in articles.slice().reverse()" :key="article.id" class="article">
        <img :src="getImgUrl(article.image)" alt="article.title" class="article__image" />
        <div class="article__content">
          <p class="article__author">{{ "By " + article.author }}</p>
          <h2 class="article__title">{{ article.title }}</h2>
          <p class="article__text">{{ summaryText(article.content, 130) }}</p>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
  import { type IArticle } from '@/core/models/db/IArticle';

  export default {
    name: 'ArticlesSection',
    setup() {
      const getImgUrl = (pic: string): string => {
        return new URL("../../../assets/images/", import.meta.url).href + `/${pic}`
      }

      return { getImgUrl }
    },
    data() {
      return{
        articles: [] as IArticle[]
      }
    },
    methods: {
      fetchArticles() {
        fetch('http://localhost:3001/articles')
          .then(response => response.json())
          .then(data => {
            this.articles = data.slice(-4)
          })
      },
      summaryText(text: string, maxLength: number): string {
        const trimmedString = text.slice(0, maxLength)

        return trimmedString.slice(0, Math.min(trimmedString.length, trimmedString.lastIndexOf(' '))) + '...'
      }
    },
    mounted() {
        this.fetchArticles()
    },
  }
</script>

<style scoped>
  .article__section {
    display: flex;
    flex-direction: column;
    margin: 40px 0;
  }

  .article__section__title {
    font-size: 2em;
    font-weight: 400;
    margin: 20px 30px;
    color: var(--color-heading);
    text-align: center;
  }

  .articles__area {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    padding: 0 30px;
  } 

  .article {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin: 10px 0;
    background-color: var(--color-background);
    border-radius: 5px;
    width: 315px;
  }

  .article__image {
    width: 100%;
    height: 200px;
    object-fit: cover;
    border-radius: 5px 5px 0 0;
  }

  .article__content {
    padding: 0 20px;
    margin: 40px 0;
  }

  .article__author {
    font-size: 0.7em;
    margin-bottom: 10px;
  }

  .article__title {
    font-size: 1.1em;
    letter-spacing: -1px;
    line-height: 1.2em;
    margin-bottom: 10px;
    font-weight: 400;
    color: var(--color-heading);
  }

  .article__text {
    font-size: 0.8em;
    line-height: 1.4em;
    color: var(--color-text);
  }

</style>