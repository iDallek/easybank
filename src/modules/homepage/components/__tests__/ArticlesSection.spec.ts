import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest';
import ArticlesSection from '../ArticlesSection.vue'

describe('ArticlesSection', () => {
  it('ensure fetches articles are rendered', async () => {
    const articles = [
      {
        id: 1,
        title: 'Article 1',
        author: 'Author 1',
        content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
        image: 'article1.jpg'
      },
      {
        id: 2,
        title: 'Article 2',
        author: 'Author 2',
        content: 'Sed ut perspiciatis unde omnis iste natus error sit voluptatem.',
        image: 'article2.jpg'
      },
      {
        id: 3,
        title: 'Article 3',
        author: 'Author 3',
        content: 'At vero eos et accusamus et iusto odio dignissimos ducimus.',
        image: 'article3.jpg'
      },
      {
        id: 4,
        title: 'Article 4',
        author: 'Author 4',
        content: 'Eaque ipsa quae ab illo inventore veritatis et quasi architecto.',
        image: 'article4.jpg'
      }
    ]
    
    vi.spyOn(window, 'fetch').mockImplementationOnce(() => {
      return Promise.resolve({
        json: () => Promise.resolve(articles)
      } as Response)
    })

    const wrapper = mount(ArticlesSection)

    setTimeout(() => {
      expect(wrapper.findAll('.article')).toHaveLength(4)
      expect(wrapper.find('.article__title').text()).toBe('Article 4')
      expect(wrapper.find('.article__author').text()).toBe('By Author 4')
      expect(wrapper.find('.article__text').text()).toBe('Eaque ipsa quae ab illo inventore veritatis et...')
      expect(wrapper.find('.article__image').attributes('src')).toBe('http://localhost/assets/images/article4.jpg')
    }, 1000)
  })
})