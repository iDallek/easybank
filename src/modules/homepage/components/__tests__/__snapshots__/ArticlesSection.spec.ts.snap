// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticlesSection > Should component render correctly 1`] = `
"<div data-v-1c7763af=\\"\\" class=\\"article__section\\">
  <h1 data-v-1c7763af=\\"\\" class=\\"article__section__title\\">Lastest Article</h1>
  <div data-v-1c7763af=\\"\\" class=\\"articles__area\\"></div>
</div>"
`;

exports[`ArticlesSection > should component render correctly 1`] = `
"<div data-v-1c7763af=\\"\\" class=\\"article__section\\">
  <h1 data-v-1c7763af=\\"\\" class=\\"article__section__title\\">Lastest Article</h1>
  <div data-v-1c7763af=\\"\\" class=\\"articles__area\\"></div>
</div>"
`;
